### PR TITLE
URL-encodes the `userInfo` component of an URI

### LINKED ISSUE #252: `Uri::withUserInfo` method should URL-encode the arguments
_This aspect is not covered the PSR-7 spec (i.e. whether an implementation should URL-encode the "user info" component)._

When I use a `userInfo` containing special characters, I expect that the `Uri::withUserInfo` method produces a valid URI. That's not the case:

```php
use GuzzleHttp\Psr7\Uri;

$uri = (new Uri('http://domain.com'))->withUserInfo('foo@bar.com', 'pass#word');
echo (string) $uri;

// Actual:    http://foo@bar.com:pass#word@domain.com
// Expected:  http://foo%40bar.com:pass%23word@domain.com
```

If I want to have a valid URI, I must call `rawurlencode` on the arguments:

```php
$uri = (new Uri('http://domain.com'))
    ->withUserInfo(rawurlencode('foo@bar.com'), rawurlencode('pass#word'));
```

I think the URL-encoding should be handled by the `Uri::withUserInfo` method, not the library consumer.

### PR DESCRIPTION
**Pull request by cedx**
Fixes #252

### CODE REVIEW

File: src/Uri.php [hunk 1]
```diff
@@ -437,9 +437,9 @@ public function withScheme($scheme)
 
     public function withUserInfo($user, $password = null)
     {
-        $info = $user;
+        $info = rawurlencode($user);
         if ($password != '') {
```
Comment (Tobion) [inline src/Uri.php]: please change that to test against `null`. an empty string should be possible and add the seperator according to rfc 3986

File: src/Uri.php [hunk 2]
```diff
@@ -437,9 +437,9 @@ public function withScheme($scheme)
 
     public function withUserInfo($user, $password = null)
     {
-        $info = $user;
+        $info = rawurlencode($user);
```
Comment (Tobion) [inline src/Uri.php]: this breaks bc with existing code as it doubles escapes stuff that people passed encoded already. so we need to make sure to not double escape. this logic is already happening for the path, query etc.
Preventing double escaping makes it ambigous when using usernames/passwords that resemble percent-encoded values like `pass%23word`. So for the generic case you still need to escape manually before calling the function. But at least this does not break bc and is consistent with path encoding handling.

Comment (Tobion) [reply src/Uri.php]: see also https://github.com/zendframework/zend-diactoros/blob/0c108ccdf5a1db0d976c13cb9140775dc828c205/src/Uri.php#L561-L570

File: src/Uri.php (hunk 1 above)
Comment (cedx) [reply src/Uri.php]: OK

File: src/Uri.php (hunk 2 above)
Comment (cedx) [reply src/Uri.php]: OK

File: src/Uri.php [hunk 3]
```diff
@@ -576,6 +576,26 @@ private function filterScheme($scheme)
...
     }
 
+    /**
+     * @param string $component
+     *
+     * @return string
+     *
+     * @throws \InvalidArgumentException If the user info is invalid.
+     */
+    private function filterUserInfoComponent($component)
```
Comment (Tobion) [inline src/Uri.php]: please also call this method from `applyParts` and add a test of this when the user info is set from the uri constructor

Comment (cedx) [reply src/Uri.php]: There's already tests for `applyParts`. Should I modify them to include special characters in `userInfo`, or create new tests?

**Comment by Tobion** [comment]
Thank you  @cedx

**Pull request merged**